lib/create-contentful-api.ts

```diff
@@ -30,118 +30,75 @@ import { FieldsType } from './types/query/util'
 import normalizeSelect from './utils/normalize-select'
 import resolveCircular from './utils/resolve-circular'
 import validateTimestamp from './utils/validate-timestamp'
-import { ChainOption, ChainOptions } from './utils/client-helpers'
+import {
+  AddChainModifier,
+  ChainModifiers,
+  ChainOptions,
+  ModifiersFromOptions,
+} from './utils/client-helpers'
 import { validateLocaleParam, validateResolveLinksParam } from './utils/validate-params'
 
 const ASSET_KEY_MAX_LIFETIME = 48 * 60 * 60
 
-export type ClientWithLinkResolutionAndWithUnresolvableLinks = BaseClient &
-  BaseClientWithAssets & {
-    withAllLocales: ClientWithAllLocalesAndWithLinkResolutionAndWithUnresolvableLinks
-    withoutLinkResolution: ClientWithoutLinkResolution
-    withoutUnresolvableLinks: ClientWithLinkResolutionAndWithoutUnresolvableLinks
-
-    getEntry<Fields extends FieldsType>(
-      id: string,
-      query?: EntryQueries
-    ): Promise<Entry<Fields, undefined>>
-
-    getEntries<Fields extends FieldsType>(
-      query?: EntriesQueries<Fields>
-    ): Promise<EntryCollection<Fields, undefined>>
-
-    // TODO: think about using collection generic as response type:
-    // ): Promise<Collection<EntryWithLinkResolution<Fields>>>
-
-    parseEntries<Fields extends FieldsType = FieldsType>(
-      data: EntryCollection<Fields, 'WITHOUT_LINK_RESOLUTION'>
-    ): EntryCollection<Fields, undefined>
-  }
+type ClientMethodsWithAllLocales<Modifiers extends ChainModifiers> = {
```

Comment:
Since I can’t dynamically add a generic for locales depending on whether the `withAllLocales` has been used, I kept two separate definitions for clients with and without `withAllLocales` internally but it’s only used here and not exported.